voila/handler.py

```diff
@@ -51,7 +51,20 @@ def get(self, path=None):
             raise tornado.web.HTTPError(404, 'file not found')
 
         # Fetch kernel name from the notebook metadata
-        kernel_name = notebook.metadata.get('kernelspec', {}).get('name', self.kernel_manager.default_kernel_name)
+        kernelspec = notebook.metadata.get('kernelspec', {})
+        kernel_name = kernelspec.get('name', self.kernel_manager.default_kernel_name)
+
+        # Find a spec matching the language if the kernel name does not exist in the kernelspecs
+        all_kernel_specs = yield tornado.gen.maybe_future(self.kernel_spec_manager.get_all_specs())
```

Comment:
Why is `yield tornado.gen.maybe_future` needed? The current implementation gives a plain dict, or can other implementations return a future?

Reply:
The kernel spec manager could be a `GatewayKernelSpecManager`, in which case the `maybe_future` would be needed: https://github.com/jupyter/notebook/blob/e498de6775b3de57f5ff827e562c179b17893064/notebook/gateway/managers.py#L521